include/p4mlir/Dialect/P4HIR/P4HIR_Ops.td

```diff
@@ -1526,6 +1529,80 @@ def SwitchOp : P4HIR_Op<"switch",
   }];
 }
 
+def ForOp : P4HIR_Op<"for",
+      [AutomaticAllocationScope,
+       DeclareOpInterfaceMethods<LoopLikeOpInterface>,
+       DeclareOpInterfaceMethods<RegionBranchOpInterface>,
+       RecursiveMemoryEffects, NoRegionArguments,
+      ]> {
+
+  let summary = "For operation";
+  let description = [{
+    The `p4hir.for` operation can execute a statement zero or more times.
+
+    Example (summing integers from 0 to 9):
+
+    ```mlir
+    %sum = p4hir.variable ["sum", init] : <!b32i>
+    p4hir.scope {
+      %i = p4hir.variable ["i", init] : <!b32i>
+      p4hir.for : cond {
+        p4hir.assign %init_val, %i : <!b32i>
+        %i_val = p4hir.read %i : <!b32i1>
+        %cond = p4hir.cmp(lt, %i_val, %limit) : !b32i, !p4hir.bool
+        p4hir.condition %cond : !p4hir.bool
+      } body {
+        %sum_val = p4hir.read %sum : <!b32i>
+        %i_val = p4hir.read %i : <!b32i>
+        %add = p4hir.binop(add, %sum_val, %i_val) : !b32i
+        p4hir.assign %add, %sum : <!b32i>
+      } updates {
+        %i_val = p4hir.read %i : <!b32i>
+        %add = p4hir.binop(add, %i_val, %one) : !b32i
+        p4hir.assign %add, %i : <!b32i>
+      }
+    }
+    ```
+  }];
+
+  let regions = (region SizedRegion<1>:$cond,
+                        MinSizedRegion<1>:$body,
+                        SizedRegion<1>:$updates);
+
+  let hasVerifier = 1;
+
+  let skipDefaultBuilders = 1;
+  let builders = [
+    OpBuilder<(ins "llvm::function_ref<void(mlir::OpBuilder &, mlir::Location)>":$condBuilder,
+                   "llvm::function_ref<void(mlir::OpBuilder &, mlir::Location)>":$bodyBuilder,
+                   "llvm::function_ref<void(mlir::OpBuilder &, mlir::Location)>":$updatesBuilder)>
+  ];
+
+  let assemblyFormat = [{
+    `:` `cond` $cond `body` $body `updates` $updates attr-dict
```

Comment:
you may also want to support annotations for body these days. See how it is done for `IfOp.